File: benchmark.py
#!/usr/bin/env python3
import time
from qubit_computer import QubitComputer, Circuit, algorithms

def benchmark_operation(name, func, iterations=1):
    """Benchmark an operation."""
    times = []
    for _ in range(iterations):
        start = time.perf_counter()
        func()
        elapsed = time.perf_counter() - start
        times.append(elapsed * 1000)
    
    avg = sum(times) / len(times)
    return avg

print("=" * 70)
print("QUBIT COMPUTER PERFORMANCE BENCHMARK (Python)")
print("=" * 70)

print("\n[1] STATE VECTOR SIZE SCALING")
print("-" * 70)
for n_qubits in [4, 8, 12, 16, 18, 20]:
    try:
        def test():
            circuit = Circuit(n_qubits)
            circuit.h(0)
            for i in range(1, n_qubits):
                circuit.cx(0, i)
            computer = QubitComputer(n_qubits)
            result = computer.run(circuit, shots=1)
        
        avg = benchmark_operation(f"GHZ({n_qubits}q)", test, iterations=3)
        state_size = 2**n_qubits
        amp_bytes = state_size * 16
        print(f"GHZ({n_qubits:2d}q) {avg:8.2f}ms | {state_size:7d} amps | {amp_bytes/1024/1024:6.1f}MB")
    except Exception as e:
        print(f"GHZ({n_qubits:2d}q) - Error: {str(e)[:50]}")

print("\n[2] GATE OPERATION OVERHEAD (8 qubits)")
print("-" * 70)
for n_gates in [10, 50, 100, 200]:
    def test():
        circuit = Circuit(8)
        for _ in range(n_gates):
            circuit.h(0)
            circuit.cx(0, 1)
            circuit.rz(0.5, 2)
        computer = QubitComputer(8)
        result = computer.run(circuit, shots=1)
    
    avg = benchmark_operation(f"Chain({n_gates:3d} gates)", test, iterations=3)
    overhead = avg / n_gates if n_gates > 0 else 0
    print(f"Chain({n_gates:3d} gates) {avg:8.2f}ms | {overhead:6.3f}ms/gate")

print("\n[3] ALGORITHM PERFORMANCE")
print("-" * 70)

def test_bell():
    circuit = algorithms.bell()
    computer = QubitComputer(2)
    result = computer.run(circuit, shots=1000)

avg = benchmark_operation("Bell (2q, 1000 shots)", test_bell, iterations=5)
print(f"Bell (2q, 1000 shots)    {avg:8.2f}ms")

def test_grover3():
    circuit = algorithms.grover(3, '011')
    computer = QubitComputer(3)
    result = computer.run(circuit, shots=1)

avg = benchmark_operation("Grover (3q)", test_grover3, iterations=5)
print(f"Grover (3q)              {avg:8.2f}ms")

def test_grover4():
    circuit = algorithms.grover(4, '0110')
    computer = QubitComputer(4)
    result = computer.run(circuit, shots=1)

avg = benchmark_operation("Grover (4q)", test_grover4, iterations=3)
print(f"Grover (4q)              {avg:8.2f}ms")

def test_grover8():
    circuit = algorithms.grover(8, '01101010')
    computer = QubitComputer(8)
    result = computer.run(circuit, shots=1)

avg = benchmark_operation("Grover (8q)", test_grover8, iterations=2)
print(f"Grover (8q)              {avg:8.2f}ms")

def test_qft8():
    circuit = algorithms.qft(8)
    computer = QubitComputer(8)
    result = computer.run(circuit, shots=1)

avg = benchmark_operation("QFT (8q)", test_qft8, iterations=3)
print(f"QFT (8q)                 {avg:8.2f}ms")

def test_qft12():
    circuit = algorithms.qft(12)
    computer = QubitComputer(12)
    result = computer.run(circuit, shots=1)

avg = benchmark_operation("QFT (12q)", test_qft12, iterations=2)
print(f"QFT (12q)                {avg:8.2f}ms")

print("\n[4] MEASUREMENT & READOUT")
print("-" * 70)
for n_qubits in [8, 12, 16]:
    def test():
        circuit = Circuit(n_qubits)
        for q in range(n_qubits):
            circuit.h(q)
        computer = QubitComputer(n_qubits)
        result = computer.run(circuit, shots=1000)
    
    avg = benchmark_operation(f"Measure({n_qubits:2d}q, 1000 shots)", test, iterations=3)
    per_shot = avg / 1000
    print(f"Measure({n_qubits:2d}q, 1000 shots) {avg:8.2f}ms | {per_shot*1000:.3f}µs/shot")

print("\n" + "=" * 70)

File: app/__tests__/benchmark.ts
import { QubitComputer, Circuit, algorithms } from '../src/core/index';

describe('Performance Benchmark', () => {
  const benchmark = async (name: string, func: () => void, iterations = 1): Promise<number> => {
    const times: number[] = [];
    for (let i = 0; i < iterations; i++) {
      const start = performance.now();
      func();
      const elapsed = performance.now() - start;
      times.push(elapsed);
    }
    return times.reduce((a, b) => a + b) / times.length;
  };

  test('state vector scaling and algorithms', async () => {
    console.log('='.repeat(70));
    console.log('QUBIT COMPUTER PERFORMANCE BENCHMARK (TypeScript)');
    console.log('='.repeat(70));

    console.log('\n[1] STATE VECTOR SIZE SCALING');
    console.log('-'.repeat(70));
    for (const n_qubits of [4, 8, 12, 16, 18, 20]) {
      const test = () => {
        const circuit = new Circuit(n_qubits);
        circuit.h(0);
        for (let i = 1; i < n_qubits; i++) {
          circuit.cx(0, i);
        }
        const computer = new QubitComputer(n_qubits);
        computer.run(circuit, 1);
      };
      
      const avg = await benchmark(`GHZ(${n_qubits}q)`, test, 3);
      const state_size = 2**n_qubits;
      const amp_bytes = state_size * 16;
      const size_mb = amp_bytes / 1024 / 1024;
      console.log(`GHZ(${String(n_qubits).padStart(2)}) ${avg.toFixed(2).padStart(8)}ms | ${state_size.toString().padStart(7)} amps | ${size_mb.toFixed(1).padStart(6)}MB`);
    }

    console.log('\n[2] GATE OPERATION OVERHEAD (8 qubits)');
    console.log('-'.repeat(70));
    for (const n_gates of [10, 50, 100, 200]) {
      const test = () => {
        const circuit = new Circuit(8);
        for (let i = 0; i < n_gates; i++) {
          circuit.h(0);
          circuit.cx(0, 1);
          circuit.rz(0.5, 2);
        }
        const computer = new QubitComputer(8);
        computer.run(circuit, 1);
      };
      
      const avg = await benchmark(`Chain(${n_gates} gates)`, test, 3);
      const overhead = n_gates > 0 ? avg / n_gates : 0;
      console.log(`Chain(${String(n_gates).padStart(3)} gates) ${avg.toFixed(2).padStart(8)}ms | ${overhead.toFixed(3).padStart(6)}ms/gate`);
    }

    console.log('\n[3] ALGORITHM PERFORMANCE');
    console.log('-'.repeat(70));

    const test_bell = () => {
      const circuit = algorithms.bell();
      const computer = new QubitComputer(2);
      computer.run(circuit, 1000);
    };
    const avg_bell = await benchmark('Bell (2q, 1000 shots)', test_bell, 5);
    console.log(`Bell (2q, 1000 shots)    ${avg_bell.toFixed(2).padStart(8)}ms`);

    const test_grover3 = () => {
      const circuit = algorithms.grover(3, '011');
      const computer = new QubitComputer(3);
      computer.run(circuit, 1);
    };
    const avg_grover3 = await benchmark('Grover (3q)', test_grover3, 5);
    console.log(`Grover (3q)              ${avg_grover3.toFixed(2).padStart(8)}ms`);

    const test_qft8 = () => {
      const circuit = algorithms.qft(8);
      const computer = new QubitComputer(8);
      computer.run(circuit, 1);
    };
    const avg_qft8 = await benchmark('QFT (8q)', test_qft8, 3);
    console.log(`QFT (8q)                 ${avg_qft8.toFixed(2).padStart(8)}ms`);

    const test_qft12 = () => {
      const circuit = algorithms.qft(12);
      const computer = new QubitComputer(12);
      computer.run(circuit, 1);
    };
    const avg_qft12 = await benchmark('QFT (12q)', test_qft12, 2);
    console.log(`QFT (12q)                ${avg_qft12.toFixed(2).padStart(8)}ms`);

    console.log('\n[4] MEASUREMENT & READOUT');
    console.log('-'.repeat(70));
    for (const n_qubits of [8, 12, 16]) {
      const test = () => {
        const circuit = new Circuit(n_qubits);
        for (let q = 0; q < n_qubits; q++) {
          circuit.h(q);
        }
        const computer = new QubitComputer(n_qubits);
        computer.run(circuit, 1000);
      };
      
      const avg = await benchmark(`Measure(${n_qubits}q)`, test, 3);
      const per_shot = avg / 1000;
      console.log(`Measure(${String(n_qubits).padStart(2)}q, 1000 shots) ${avg.toFixed(2).padStart(8)}ms | ${(per_shot*1000).toFixed(3).padStart(7)}µs/shot`);
    }

    console.log('\n' + '='.repeat(70));
  }, 120000);
});
